File: app03_csharp_nft/backend/NftCerts/Api/CertificateDtoMapper.cs
using NftCerts.Certificates;
using NftCerts.Config;

namespace NftCerts.Api;

// Maps a Certificate to the API DTO shared with app01/app02, including marketplace deep links.
public class CertificateDtoMapper(ExplorerLinkProperties explorer)
{
    public object ToDto(Certificate certificate) => new
    {
        tokenId = certificate.TokenId,
        artworkId = certificate.Artwork.Id,
        title = certificate.Artwork.Title,
        contentHashHex = certificate.ContentHashHex,
        contractAddress = certificate.ContractAddress,
        txHash = certificate.TxHash,
        ownerAddress = certificate.OwnerAddress,
        royaltyPercentageBps = certificate.RoyaltyPercentageBps,
        imageIpfsUri = certificate.Artwork.ImageIpfsUri,
        metadataIpfsUri = certificate.Artwork.MetadataIpfsUri,
        etherscanUrl = BuildEtherscanUrl(certificate),
        openSeaUrl = BuildOpenSeaUrl(certificate),
        raribleUrl = BuildRaribleUrl(certificate),
    };

    public string BuildEtherscanUrl(Certificate certificate) =>
        explorer.EtherscanBaseUrl + certificate.TxHash;

    public string BuildOpenSeaUrl(Certificate certificate) =>
        $"{explorer.OpenseaBaseUrl}{certificate.ContractAddress}/{certificate.TokenId}";

    public string BuildRaribleUrl(Certificate certificate) =>
        $"{explorer.RaribleBaseUrl}{certificate.ContractAddress}:{certificate.TokenId}";
}

File: app02_cpp_nft/backend/src/api/CertificateDtoMapper.cpp
#include "api/CertificateDtoMapper.h"

namespace nftcerts::api {

Json::Value CertificateDtoMapper::toDto(const certificate::Certificate& certificate) const {
    Json::Value dto;
    dto["tokenId"] = Json::Value(static_cast<Json::UInt64>(certificate.tokenId));
    dto["artworkId"] = certificate.artwork.id;
    dto["title"] = certificate.artwork.title;
    dto["contentHashHex"] = certificate.contentHashHex;
    dto["contractAddress"] = certificate.contractAddress;
    dto["txHash"] = certificate.txHash;
    dto["ownerAddress"] = certificate.ownerAddress;
    dto["royaltyPercentageBps"] = certificate.royaltyPercentageBps;
    dto["imageIpfsUri"] = certificate.artwork.imageIpfsUri;
    dto["metadataIpfsUri"] = certificate.artwork.metadataIpfsUri;
    dto["etherscanUrl"] = buildEtherscanUrl(certificate);
    dto["openSeaUrl"] = buildOpenSeaUrl(certificate);
    dto["raribleUrl"] = buildRaribleUrl(certificate);
    return dto;
}

std::string CertificateDtoMapper::buildEtherscanUrl(const certificate::Certificate& certificate) const {
    return explorerLinkProperties_.etherscanBaseUrl + certificate.txHash;
}

std::string CertificateDtoMapper::buildOpenSeaUrl(const certificate::Certificate& certificate) const {
    return explorerLinkProperties_.openseaBaseUrl + certificate.contractAddress + "/" +
           std::to_string(certificate.tokenId);
}

std::string CertificateDtoMapper::buildRaribleUrl(const certificate::Certificate& certificate) const {
    return explorerLinkProperties_.raribleBaseUrl + certificate.contractAddress + ":" +
           std::to_string(certificate.tokenId);
}

}  // namespace nftcerts::api
